File: app/src/main/java/com/example/bg6tracker/KalmanTracker.java
package com.example.bg6tracker;

public class KalmanTracker {

    private float x;
    private float y;
    private float vx;
    private float vy;

    private boolean initialized = false;

    private final float q = 0.4f;
    private final float r = 1.5f;

    public synchronized void update(
            float measurementX,
            float measurementY,
            float dt) {

        if (!initialized) {
            x = measurementX;
            y = measurementY;
            vx = 0;
            vy = 0;
            initialized = true;
            return;
        }

        if (dt <= 0 || dt > 0.2f) {
            dt = 0.01f;
        }

        float predictedX = x + vx * dt;
        float predictedY = y + vy * dt;

        float errorX = measurementX - predictedX;
        float errorY = measurementY - predictedY;

        float gain = q / (q + r);

        x = predictedX + gain * errorX;
        y = predictedY + gain * errorY;

        vx = vx + gain * errorX / dt;
        vy = vy + gain * errorY / dt;
    }

    public synchronized float predict(float seconds) {
        return x + vx * seconds;
    }

    public synchronized float getX() {
        return x;
    }

    public synchronized float getY() {
        return y;
    }

    public synchronized float getVelocityX() {
        return vx;
    }

    public synchronized float getVelocityY() {
        return vy;
    }

    public synchronized boolean isInitialized() {
        return initialized;
    }
}

File: app/src/main/java/com/example/bg6tracker/KalmanTracker.kt
package com.example.bg6tracker

import kotlin.math.max

class KalmanTracker {

    /*
     * State:
     *
     * x
     * y
     * vx
     * vy
     */
    private var x =
        0.0

    private var y =
        0.0

    private var vx =
        0.0

    private var vy =
        0.0

    private var initialized =
        false

    /*
     * Measurement smoothing.
     *
     * Giá trị cao:
     * theo measurement nhanh hơn.
     *
     * Giá trị thấp:
     * mượt hơn nhưng trễ hơn.
     */
    private val alpha =
        0.65

    fun reset() {

        x = 0.0
        y = 0.0

        vx = 0.0
        vy = 0.0

        initialized =
            false
    }

    /*
     * Nhận measurement mới.
     *
     * Return:
     *
     * [x, y, vx, vy]
     */
    @Synchronized
    fun update(
        measurementX: Double,
        measurementY: Double,
        dtInput: Double
    ): DoubleArray {

        val dt =
            sanitizeDt(
                dtInput
            )

        if (!initialized) {

            x =
                measurementX

            y =
                measurementY

            vx = 0.0
            vy = 0.0

            initialized =
                true

            return state()
        }

        /*
         * Prediction.
         */
        val predictedX =
            x +
                vx * dt

        val predictedY =
            y +
                vy * dt

        /*
         * Correction.
         */
        val correctedX =
            predictedX +
                alpha *
                (
                    measurementX -
                        predictedX
                )

        val correctedY =
            predictedY +
                alpha *
                (
                    measurementY -
                        predictedY
                )

        /*
         * Velocity.
         */
        val newVx =
            (
                correctedX -
                    x
                ) / dt

        val newVy =
            (
                correctedY -
                    y
                ) / dt

        /*
         * Giới hạn velocity để tránh
         * một measurement lỗi tạo ra
         * spike quá lớn.
         */
        vx =
            newVx.coerceIn(
                -5000.0,
                5000.0
            )

        vy =
            newVy.coerceIn(
                -5000.0,
                5000.0
            )

        x =
            correctedX

        y =
            correctedY

        return state()
    }

    /*
     * Prediction khi không có measurement
     * đáng tin cậy.
     */
    @Synchronized
    fun predict(
        dtInput: Double
    ): DoubleArray {

        val dt =
            sanitizeDt(
                dtInput
            )

        if (!initialized) {
            return state()
        }

        x +=
            vx * dt

        y +=
            vy * dt

        return state()
    }

    /*
     * Predict trước một khoảng thời gian.
     *
     * Ví dụ:
     *
     * predictAhead(0.030)
     *
     * = dự đoán 30ms.
     */
    @Synchronized
    fun predictAhead(
        milliseconds: Double
    ): DoubleArray {

        if (!initialized) {
            return state()
        }

        val seconds =
            (
                milliseconds /
                    1000.0
            ).coerceIn(
                0.0,
                0.2
            )

        return doubleArrayOf(
            x + vx * seconds,
            y + vy * seconds,
            vx,
            vy
        )
    }

    fun isInitialized():
        Boolean {

        return initialized
    }

    fun getX():
        Double {

        return x
    }

    fun getY():
        Double {

        return y
    }

    fun getVelocityX():
        Double {

        return vx
    }

    fun getVelocityY():
        Double {

        return vy
    }

    private fun state():
        DoubleArray {

        return doubleArrayOf(
            x,
            y,
            vx,
            vy
        )
    }

    private fun sanitizeDt(
        dtInput: Double
    ): Double {

        return max(
            0.001,
            dtInput.coerceAtMost(
                0.2
            )
        )
    }
}
